tests/run/pure_py.py

```diff
@@ -319,6 +319,32 @@ def test_cdef_nogil(x):
     return result
 
 
+@cython.cfunc
+@cython.inline
+def has_inner_func1(x):
+    # the inner function must remain a Python function
+    # (and inline must not be applied to it)
+    def inner():
+        return x
+    return inner
+
+
+with cython.cfunc:
+    def has_inner_func2(x):
+        # cfunc is not applied to "inner"
```

Comment:
Hmm. I wonder if that is the "expected" behaviour. I fully agree that a decorator should only apply to the decorated function, but the with-statement? I think one could argue both ways here. On the one hand, it should apply to the whole block, including everything that's in there. On the other hand, it is applied as part of the control flow, and the inner functions are not being defined in that context.

I guess it's best to not inherit it. That feels the safest and should be most easy to handle for users.

Reply:
I sort of agree here - the decorator clearly only applies to the function it's decorating, but the with statement feels more general.

The main reason not to do it is that it's hard to distinguish decorator from with statement